Add tcall_dispatcher unit tests and int tests

dcristoloveanu opened the pull request.

anporumb commented on inc/clds/tcall_dispatcher_ll.h:
```diff
@@ -10,8 +10,274 @@
+        if (result == NULL)                                                                                                                                         \
+        {                                                                                                                                                           \
+            /* Codes_SRS_TCALL_DISPATCHER_01_013: [ If there are any failures then TCALL_DISPATCHER_REGISTER_TARGET(T) shall fail and return NULL. ] */             \
+            LogError("malloc(sizeof(TCALL_DISPATCHER_TARGET_TYPEDEF_NAME(T))=%zu) failed", sizeof(TCALL_DISPATCHER_TARGET_TYPEDEF_NAME(T)));                        \
```
>TCALL_DISPATCHER_TARGET_TYPEDEF_NAME [](http://example.com/codeflow?start=36&length=36)

could use a MU_TOSTRING so we get the expanded "what it is"

dcristoloveanu replied:
MU_TOSTRINGified the T, I think that gives us the best info when looking at logs

anporumb commented on inc/clds/tcall_dispatcher_ll.h:
```diff
@@ -10,8 +10,274 @@
+typedef void (*TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(T))(void* context MU_FOR_EACH_2(TCALL_DISPATCHER_ARG_IN_SIGNATURE, __VA_ARGS__)); \
+typedef struct TCALL_DISPATCHER_TARGET_STRUCT_TYPE_NAME_TAG(T) TCALL_DISPATCHER_TARGET_TYPEDEF_NAME(T);         \
+typedef struct TCALL_DISPATCHER_TARGET_STRUCT_TYPE_NAME_TAG(T)* TCALL_DISPATCHER_TARGET_HANDLE_NAME(T);         \
+struct TCALL_DISPATCHER_TARGET_STRUCT_TYPE_NAME_TAG(T)                                                          \
```
>TCALL_DISPATCHER_TARGET_STRUCT_TYPE_NAME_TAG [](http://example.com/codeflow?start=7&length=44)

looking at `tcall_dispatcher.h` and this header, I think the user would get this struct as a complete type, that is, anybody using any kind of TCALL_DISPATCHER would see the fields of this struct, but they NEVER should do anything with them.

Any chance TCALL_DISPATCHER_TARGET_STRUCT_TYPE_NAME_TAG can be turned into an incomplete type for the user?

dcristoloveanu replied:
Yes, ideally. I tried and it proved more than 10 minutes, so it shall be done "when time comes".

anporumb commented:
> `TCALL_DISPATCHER_TYPE_DECLARE(T)` is a macro to be used in a .c file to define all the needed functions for `TCALL_DISPATCHER(T)`.

Would be nice to have the same example as in line 83 listed for the .c file too, and spec that the `...` arguments are the same in both cases.

---
Refers to: devdoc/tcall_dispatcher_requirements.md:91 in 80080ae. [](commit_id = 80080ae014c700e09a5bb774ed940428371cbfae, deletion_comment = False)

anporumb commented:
> `TCALL_DISPATCHER_TYPE_DECLARE(T)` is a macro to be used in a .c file to define all the needed functions for `TCALL_DISPATCHER(T)`.

And by the way, this is DEFINE, not DECLARE, maybe copy&paste? :)

---
In reply to: [1775278774](https://github.com/Azure/clds/pull/199#issuecomment-1775278774) [](http://example.com/codeflow?ancestors=1775278774)

---
Refers to: devdoc/tcall_dispatcher_requirements.md:91 in 80080ae. [](commit_id = 80080ae014c700e09a5bb774ed940428371cbfae, deletion_comment = False)

anporumb commented on tests/CMakeLists.txt:
```diff
@@ -16,7 +16,10 @@ if(${run_unittests})
     build_test_folder(mpsc_lock_free_queue_ut)
-    build_test_folder(tcall_dispatcher_ut)
+if(WIN32)
+        # this test has a problem on Linux with the lack of srw_lock_ll reals
```
>reals [](http://example.com/codeflow?start=72&length=5)

should mention the devops task that's supposed to fix it :)

dcristoloveanu replied:
Of course, added.

anporumb commented on inc/clds/tcall_dispatcher_ll.h:
```diff
@@ -10,8 +10,274 @@
 #include <inttypes.h>
```
>inttypes [](http://example.com/codeflow?start=10&length=8)

not immediately clear why this is needed.

dcristoloveanu replied:
Not needed, removed

anporumb commented on tests/tcall_dispatcher_int/tcall_dispatcher_int.c:
```diff
@@ -0,0 +1,261 @@
+
+// This is the call dispatcher used for the tests
+TCALL_DISPATCHER_DEFINE_CALL_TYPE(FOO, int32_t, x);
+THANDLE_TYPE_DECLARE(TCALL_DISPATCHER_TYPEDEF_NAME(FOO))
```
>THANDLE_TYPE_DECLARE [](http://example.com/codeflow?start=0&length=20)

Why is a THANDLE needed here, I would expect that everything is handled through TCALLs...

anporumb replied:
Same for line 24.

anporumb replied:
oooh, there are "hints" in the macros... wondering if we can do without them. 

They sound very familiar with theones we have for TARRAY...

dcristoloveanu replied:
They are extremely like the TARRAY ones, got inspired by TARRAY.

In theory we could get rid of them. I'll do a task.

dcristoloveanu replied:
[https://msazure.visualstudio.com/One/_sprints/taskboard/Azure%20Messaging%20Store/One/Custom/AzureMessaging/Germanium/GeM1?workitem=25576397](https://msazure.visualstudio.com/One/_sprints/taskboard/Azure%20Messaging%20Store/One/Custom/AzureMessaging/Germanium/GeM1?workitem=25576397)

anporumb commented on tests/tcall_dispatcher_int/tcall_dispatcher_int.c:
```diff
@@ -0,0 +1,261 @@
+THANDLE_TYPE_DECLARE(TCALL_DISPATCHER_TYPEDEF_NAME(FOO))
+TCALL_DISPATCHER_TYPE_DECLARE(FOO, int32_t, x);
+
+THANDLE_TYPE_DEFINE(TCALL_DISPATCHER_TYPEDEF_NAME(FOO));
```
The introduction of this type for the purpose of int tests should be hosted in its own files: a .h and a .c file separated from the int tests - this is most likely to emulate the real case where these TCALLs are used.

dcristoloveanu replied:
Separated hosting for one of these (the FOO).

anporumb commented on tests/tcall_dispatcher_int/tcall_dispatcher_int.c:
```diff
@@ -0,0 +1,261 @@
+// This is the call dispatcher used for the tests
+TCALL_DISPATCHER_DEFINE_CALL_TYPE(FOO, int32_t, x);
+THANDLE_TYPE_DECLARE(TCALL_DISPATCHER_TYPEDEF_NAME(FOO))
+TCALL_DISPATCHER_TYPE_DECLARE(FOO, int32_t, x);
```
>TCALL_DISPATCHER_TYPE_DECLARE [](http://example.com/codeflow?start=0&length=29)

Int tests should have some unit where a function that is `void f(void)` can be called.

dcristoloveanu replied:
Added the no args and 3 args int tests

anporumb commented on tests/tcall_dispatcher_int/tcall_dispatcher_int.c:
```diff
@@ -0,0 +1,261 @@
+THANDLE_TYPE_DECLARE(TCALL_DISPATCHER_TYPEDEF_NAME(FOO))
+TCALL_DISPATCHER_TYPE_DECLARE(FOO, int32_t, x);
+
+THANDLE_TYPE_DEFINE(TCALL_DISPATCHER_TYPEDEF_NAME(FOO));
```
>FOO [](http://example.com/codeflow?start=50&length=3)

general: int tests should have a test where 2 of these TCALLs are hosted in the same module. Just to make sure we avoid any potential linker doubly defined symbols.

dcristoloveanu replied:
Added now, we have them also hosted in UTs. They are happy to be hosted together.

anporumb commented on inc/clds/tcall_dispatcher_ll.h:
```diff
@@ -10,8 +10,274 @@
+    else                                                                                                                                                            \
+    {                                                                                                                                                               \
+        /* Codes_SRS_TCALL_DISPATCHER_01_002: [ TCALL_DISPATCHER_CREATE(T) shall call srw_lock_ll_init to initialize the lock used by the TCALL_DISPATCHER instance. ]*/ \
+        if (srw_lock_ll_init(&result->lock) != 0)                                                                                                                   \
```
>srw_lock_ll_init [](http://example.com/codeflow?start=12&length=16)

what has `_init` should have `_deinit` someplace and I cannot find it. Reviewer code blindness?

dcristoloveanu replied:
Hah, that was a miss! Nice spotting it.

anporumb commented on inc/clds/tcall_dispatcher_ll.h:
```diff
@@ -10,8 +10,274 @@
+    }                                                                                                                           \
+    else                                                                                                                        \
+    {                                                                                                                           \
+    }                                                                                                                           \
```
so what happens if at TCALL_ASSIGN(...)(H, NULL) there are some elements still allocated in the doubly_linked_list?

dcristoloveanu replied:
Specs added, tests added, code has been filled in. Cheers!

anporumb commented on tests/tcall_dispatcher_int/tcall_dispatcher_int.c:
```diff
@@ -0,0 +1,261 @@
+        }
+    } while (1);
+
+    free((void*)test_context.call_target_handles);
```
>void [](http://example.com/codeflow?start=10&length=4)

why is this cast necessary?

dcristoloveanu replied:
Because of the volatile_atomic TCALL_DISPATCHER_TARGET_HANDLE(FOO)*.
I knew you will notice this, I was expecting it!

anporumb commented on tests/tcall_dispatcher_ut/tcall_dispatcher_ut.c:
```diff
@@ -62,4 +112,412 @@ TEST_FUNCTION_CLEANUP(method_cleanup)
+TEST_FUNCTION(TCALL_DISPATCHER_DISPATCH_CALL_with_NULL_tcall_dispatcher_fails)
+{
+    // arrange
+    TCALL_DISPATCHER(FOO) call_dispatcher = test_create_tcall_dispatcher();
```
>call_dispatcher [](http://example.com/codeflow?start=26&length=15)

not really needed, is it...

dcristoloveanu replied:
Removed

anporumb commented on tests/tcall_dispatcher_ut/tcall_dispatcher_ut.c:
```diff
@@ -62,4 +112,412 @@ TEST_FUNCTION_CLEANUP(method_cleanup)
+
+    STRICT_EXPECTED_CALL(srw_lock_ll_acquire_shared(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(test_target_1((void*)0x4242, 41));
+    STRICT_EXPECTED_CALL(test_target_2(NULL, 41));
```
>NULL [](http://example.com/codeflow?start=39&length=4)

nice

dcristoloveanu replied:
:-)

dcristoloveanu commented on tests/tcall_dispatcher_ut/tcall_dispatcher_ut.c:
```diff
@@ -62,4 +112,412 @@ TEST_FUNCTION_CLEANUP(method_cleanup)
+TEST_FUNCTION(TCALL_DISPATCHER_DISPATCH_CALL_calls_a_target_with_3_args)
+{
+    // arrange
+    const char* test_string = "muma padurilor";
```
:-)

dcristoloveanu commented:
> `TCALL_DISPATCHER_TYPE_DECLARE(T)` is a macro to be used in a .c file to define all the needed functions for `TCALL_DISPATCHER(T)`.

Fixed the copy/pasta and added the example.

---
In reply to: [1775279601](https://github.com/Azure/clds/pull/199#issuecomment-1775279601) [](http://example.com/codeflow?ancestors=1775279601,1775278774)

---
Refers to: devdoc/tcall_dispatcher_requirements.md:91 in 80080ae. [](commit_id = 80080ae014c700e09a5bb774ed940428371cbfae, deletion_comment = False)

anporumb reviewed (approved):
:shipit:
